Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp

```diff
@@ -0,0 +1,386 @@
+    }
+}
+
+Ref<AudioEncoder::EncodePromise> InternalAudioEncoderCocoa::encode(AudioEncoder::RawFrame&& rawFrame)
+{
+    assertIsCurrent(queueSingleton());
+
+    RetainPtr cmSample = downcast<PlatformRawAudioDataCocoa>(rawFrame.frame)->sampleBuffer();
+    ASSERT(cmSample);
+    if (auto error = PAL::CMSampleBufferSetOutputPresentationTimeStamp(cmSample.get(), PAL::CMTimeMake(rawFrame.timestamp, 1000000)))
+        RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferConverter CMSampleBufferSetOutputPresentationTimeStamp failed with %d", error);
+
+    CMFormatDescriptionRef formatDescription = PAL::CMSampleBufferGetFormatDescription(cmSample.get());
+    if (!formatDescription)
+        return EncodePromise::createAndReject("Couldn't retrieve AudioData's format description"_s);
+    const AudioStreamBasicDescription* const asbd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(formatDescription);
+    if (!asbd)
+        return EncodePromise::createAndReject("Couldn't retrieve AudioData's basic description"_s);
+
+    if (*asbd != m_inputDescription) {
+        if (m_converter)
+            converter()->finish();
+        m_inputDescription = *asbd;
+
+        AudioSampleBufferConverter::Options options = {
+            .format = m_codec,
+            .description = m_outputDescription->streamDescription(),
+            .outputBitRate = m_config.bitRate ? std::optional { m_config.bitRate } : std::nullopt,
+            .generateTimestamp = false,
+            .bitrateMode = m_config.bitRateMode,
+            .complexity = m_complexity,
+            .packetlossperc = m_packetlossperc,
+            .useinbandfec = m_useinbandfec
+        };
+        m_converter = AudioSampleBufferConverter::create(compressedAudioOutputBufferCallback, this, options);
+        if (!m_converter) {
+            RELEASE_LOG_ERROR(MediaStream, "InternalAudioEncoderCocoa::encode: creation of converter failed");
+            return EncodePromise::createAndReject("InternalAudioEncoderCocoa::encode: creation of converter failed"_s);
+        }
+    }
+
+    m_needsFlushing = true;
+
+    ASSERT(m_converter);
+    AudioEncoder::EncodePromise::Producer producer;
+    Ref promise = producer.promise();
+    converter()->addSampleBuffer(cmSample.get())->whenSettled(queueSingleton(), [producer = WTFMove(producer)](auto result) mutable {
+        if (result)
+            producer.resolve();
+        else
+            producer.reject("InternalAudioEncoderCocoa encoding failed"_s);
+    });
+    return promise;
+}
+
+Ref<GenericPromise> InternalAudioEncoderCocoa::flush()
+{
+    assertIsCurrent(queueSingleton());
+
+    if (!m_converter || !m_needsFlushing)
+        return GenericPromise::createAndResolve(); // No frame encoded yet, nothing to flush;
+    return converter()->drain()->whenSettled(queueSingleton(), [protectedThis = Ref { *this }] {
+        assertIsCurrent(queueSingleton());
+        protectedThis->processEncodedOutputs();
```

Comment:
If this fails, maybe we should fail the flush promise.

Reply:
done